libs/librepcb/common/fileio/sexpression.cpp

```diff
@@ -289,19 +302,151 @@ SExpression SExpression::createLineBreak() {
+  }
+  SExpression root = parse(contentStr, index, filePath);
+  if (index < contentStr.length()) {
+    throw FileParseError(__FILE__, __LINE__, filePath, -1, -1, QString(),
+                         "File contains more than one root node.");
+  }
+  return root;
+}
+
+/*******************************************************************************
+ *  Private Methods
+ ******************************************************************************/
+
+SExpression SExpression::parse(const QString& content, int& index,
+                               const FilePath& filePath) {
+  Q_ASSERT(index < content.length());
+
+  if (content.at(index) == '(') {
+    return parseList(content, index, filePath);
+  } else if (content.at(index) == '"') {
+    return createString(parseString(content, index, filePath));
+  } else {
+    return createToken(parseToken(content, index, filePath));
+  }
+}
+
+SExpression SExpression::parseList(const QString& content, int& index,
+                                   const FilePath& filePath) {
+  Q_ASSERT((index < content.length()) && (content.at(index) == '('));
+
+  ++index;  // consume the '('
+
+  SExpression list = createList(parseToken(content, index, filePath));
+
+  while (true) {
+    if (index >= content.length()) {
+      throw FileParseError(__FILE__, __LINE__, filePath, -1, -1, QString(),
+                           "S-Expression node ended without closing ')'.");
+    }
+    if (content.at(index) == ')') {
+      ++index;  // consume the ')'
+      skipWhitespaceAndComments(content, index);  // consume following spaces
+      break;
     } else {
+      list.appendChild(parse(content, index, filePath));
+    }
+  }
+
+  return list;
+}
+
+QString SExpression::parseToken(const QString& content, int& index,
+                                const FilePath& filePath) {
+  int oldIndex = index;
+  while ((index < content.length()) && (isValidTokenChar(content.at(index)))) {
+    ++index;
+  }
+  QString token = content.mid(oldIndex, index - oldIndex);
+  if (token.isEmpty()) {
+    throw FileParseError(
+        __FILE__, __LINE__, filePath, -1, -1, QString(),
+        QString("Invalid token character detected: '%1'")
+            .arg(index < content.length() ? content.at(index) : QChar()));
+  }
+  skipWhitespaceAndComments(content, index);  // consume following spaces
+  return token;
+}
+
+QString SExpression::parseString(const QString& content, int& index,
+                                 const FilePath& filePath) {
+  ++index;  // consume the '"'
+
+  // Note: Until LibrePCB 0.1.5 we used the sexpresso library for escaping
+  // strings. This library escaped more characters than we do now. To still
+  // support reading the file format 0.1, we have to keep support for the
+  // old escaping behavior.
+  static QHash<QChar, QChar> escapedChars = {
+      {'\'', '\''},  // Single quote
+      {'"', '"'},  // Double quote
+      {'?', '\?'},  // Question mark
+      {'\\', '\\'},  // Backslash
```

Comment:
The quotes and the backslash are the same before/after escaping. Is that on purpose?

Reply:
These characters are only replaced if they appear after a backslash, i.e. `\"` is replaced by `"`. It just looks a bit strange because characters like `\` need to be escaped within the literal, so it looks like `\\` is replaced by `\\`, but actually it replaces `\\` by `\` :)